core/block_dragger.ts

```diff
@@ -173,32 +191,178 @@ export class BlockDragger implements IBlockDragger {
    * display accordingly.
    *
    * @param e The most recent move event.
-   * @param currentDragDeltaXY How far the pointer has moved from the position
+   * @param delta How far the pointer has moved from the position
    *     at the start of the drag, in pixel units.
    */
-  drag(e: PointerEvent, currentDragDeltaXY: Coordinate) {
-    const delta = this.pixelsToWorkspaceUnits_(currentDragDeltaXY);
+  drag(e: PointerEvent, delta: Coordinate) {
+    const block = this.draggingBlock_;
+    this.moveBlock(block, delta);
+    this.updateDragTargets(e, block);
+    this.wouldDeleteBlock_ = this.wouldDeleteBlock(e, block, delta);
+    this.updateCursorDuringBlockDrag_();
+    this.updateConnectionPreview(block, delta);
+  }
+
+  private moveBlock(draggingBlock: BlockSvg, dragDelta: Coordinate) {
+    const delta = this.pixelsToWorkspaceUnits_(dragDelta);
     const newLoc = Coordinate.sum(this.startXY_, delta);
-    this.draggingBlock_.moveDuringDrag(newLoc);
+    draggingBlock.moveDuringDrag(newLoc);
+  }
 
-    const oldDragTarget = this.dragTarget_;
-    this.dragTarget_ = this.workspace_.getDragTarget(e);
+  private updateDragTargets(e: PointerEvent, draggingBlock: BlockSvg) {
+    const newDragTarget = this.workspace_.getDragTarget(e);
+    if (this.dragTarget_ !== newDragTarget) {
+      this.dragTarget_?.onDragExit(draggingBlock);
+      newDragTarget?.onDragEnter(draggingBlock);
+    }
+    newDragTarget?.onDragOver(draggingBlock);
+    this.dragTarget_ = newDragTarget;
+  }
 
-    this.draggedConnectionManager_.update(delta, this.dragTarget_);
-    const oldWouldDeleteBlock = this.wouldDeleteBlock_;
-    this.wouldDeleteBlock_ = this.draggedConnectionManager_.wouldDeleteBlock;
-    if (oldWouldDeleteBlock !== this.wouldDeleteBlock_) {
-      // Prevent unnecessary add/remove class calls.
-      this.updateCursorDuringBlockDrag_();
+  /**
+   * Returns true if we would delete the block if it was dropped at this time,
+   * false otherwise.
+   */
+  private wouldDeleteBlock(
+    e: PointerEvent,
+    draggingBlock: BlockSvg,
+    delta: Coordinate,
+  ): boolean {
+    const dragTarget = this.workspace_.getDragTarget(e);
+    if (!dragTarget) return false;
+
+    const componentManager = this.workspace_.getComponentManager();
+    const isDeleteArea = componentManager.hasCapability(
```

Comment:
It's unfortunate that delete areas are hardcoded like this.  It would be nice if there were definable target areas, that would do something on their own if something were dragged into them.  E.g. delete areas, duplication areas, backpacks, stack fragmenters...  Thus the areas would instead return "I'll be handling this block when dropped, don't suggest anything".

Reply:
Yeah we have `onDrop` but we don't actually let drag targets handle the deleting... no idea why. But the block dragger needs to handle deleting for backwards compatibility :/